### PR TITLE
Use custom queryable interfaces.

### PR DESCRIPTION
**Pull request by grokys**
Previously we were using `IQueryable` as our queryable interface. This had a few problems:

- There was no distinction between querying a list and a single value, so querying a single value had to be done like: `(await Connection.Run(query)).Single();`
- `IQueryable` pulled in a bunch of extension methods that aren't actually valid for GraphQL like `.Where()`
- We're always async but `IQueryable`s only support synchronous execution so we had to execute via `Connection.Run`

This PR replaces `IQueryable` with our own "queryable" interfaces, `IQueryableValue` and `IQueryableList` so whether the query returns a single value or an enumerable is now encoded in the type system.

### CODE REVIEW

File: Octokit.GraphQL.Core.UnitTests/QueryBuilderTests.cs [hunk 1]
```diff
@@ -84,85 +84,85 @@ public void SimpleQuery_Select_Append_Two_Identical_Members()
...
-
-            var notSupportedException = Assert.Throws<NotSupportedException>(() =>
-            {
-                expression.ToList();
-            });
-
-            Assert.Equal(notSupportedException.Message, "QueryProvider cannot be executed");
-        }
+        ////[Fact]
+        ////public void SimpleQuery_Where_Error()
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core.UnitTests/QueryBuilderTests.cs]: What do you want to do about these tests?

File: Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs [hunk 2]
```diff
@@ -51,8 +49,8 @@ public void Data_Select_Single_Member()
                 .Data
                 .Select(x => x.Id);
 
-            Expression<Func<JObject, IEnumerable<string>>> expected = data =>
-                ExpressionMethods.SelectEntity(data["data"]["data"], x => x["id"].ToObject<string>());
+            Expression<Func<JObject, string>> expected = data =>
+                Rewritten.Value.Select(data["data"]["data"], x => x["id"].ToObject<string>());
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs]: We also want to explain/understand why `ToObject` is where it is here.

File: Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs [hunk 3]
```diff
@@ -48,11 +46,11 @@ public void SimpleQuery_Select_Multiple_Members()
...
         {
             var expression = new TestQuery()
-                .Data
+                .QueryItems
                 .Select(x => x.Id);
 
-            Expression<Func<JObject, IEnumerable<string>>> expected = data =>
-                ExpressionMethods.SelectEntity(data["data"]["data"], x => x["id"].ToObject<string>());
+            Expression<Func<JObject, string>> expected = data =>
+                Rewritten.Value.Select(data["data"]["queryItems"], x => x["id"].ToObject<string>());
```
Comment (grokys) [inline Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs]: This should be `Rewritten.List.Select`.

File: Octokit.GraphQL.Core.UnitTests/QueryBuilderTests.cs (hunk 1 above)
Comment (grokys) [reply Octokit.GraphQL.Core.UnitTests/QueryBuilderTests.cs]: I think we can remove these tests now, the invalid methods aren't even available.

File: Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs (hunk 2 above)
Comment (grokys) [reply Octokit.GraphQL.Core.UnitTests/ExpressionRewiterTests.cs]: Ok, I know why this is now. The final `ToObject<string>` is added [right at the end, after the expression has been rewritten

**Pull request merged**